lib/ansible/modules/cloud/amazon/ec2_eni.py

```diff
@@ -378,161 +534,238 @@ def modify_eni(connection, vpc_id, module, eni):
-                eni.detach(force_detach)
+            if "Attachment" in eni:
+                connection.detach_network_interface(
+                    AttachmentId=eni["Attachment"]["AttachmentId"],
+                    Force=True
+                )
                 # Wait to allow detachment to finish
-                wait_for_eni(eni, "detached")
-                eni.update()
-            eni.delete()
+                wait_for(correct_status, connection, "detached", module)
+            connection.delete_network_interface(NetworkInterfaceId=eni_id)
             changed = True
         else:
-            eni.delete()
+            connection.delete_network_interface(NetworkInterfaceId=eni_id)
             changed = True
 
         module.exit_json(changed=changed)
-    except BotoServerError as e:
+    except botocore.exceptions.ClientError as e:
         regex = re.compile('The networkInterface ID \'.*\' does not exist')
         if regex.search(e.message) is not None:
             module.exit_json(changed=False)
         else:
-            module.fail_json(msg=e.message)
+            module.fail_json_aws(e, "Failure during delete of {0}".format(eni_id))
 
 
-def detach_eni(eni, module):
+def detach_eni(connection, eni, module):
 
     attached = module.params.get("attached")
 
     force_detach = module.params.get("force_detach")
-    if eni.attachment is not None:
-        eni.detach(force_detach)
-        wait_for_eni(eni, "detached")
+    if "Attachment" in eni:
+        connection.detach_network_interface(
+            AttachmentId=eni["Attachment"]["AttachmentId"],
+            Force=force_detach
+        )
+        wait_for(correct_status, connection, "detached", module)
         if attached:
             return
-        eni.update()
+        eni = uniquely_find_eni(connection, module)
         module.exit_json(changed=True, interface=get_eni_info(eni))
     else:
         module.exit_json(changed=False, interface=get_eni_info(eni))
 
 
-def uniquely_find_eni(connection, module):
+def uniquely_find_eni(connection, module, eni=None):
+
+    if eni:
+        # In the case of create, eni_id will not be a param but we can still get the eni_id after creation
+        if "NetworkInterfaceId" in eni:
+            eni_id = eni["NetworkInterfaceId"]
+        else:
+            eni_id = None
+    else:
+        eni_id = module.params.get("eni_id")
 
-    eni_id = module.params.get("eni_id")
     private_ip_address = module.params.get('private_ip_address')
     subnet_id = module.params.get('subnet_id')
     instance_id = module.params.get('instance_id')
     device_index = module.params.get('device_index')
     attached = module.params.get('attached')
+    name = module.params.get("name")
+
+    filters = []
+
+    # proceed only if we're unequivocally specifying an ENI
+    if eni_id is None and private_ip_address is None and (instance_id is None and device_index is None):
+        return None
+
+    filter_set = False
```

Comment:
filter_set seems redundant - all of the "if" statements below can just check filters directly.

Reply:
@mike-healey it looks like that boolean is used to determine if the function should return `None` or not. However, it looks like one could equally accomplish that by just checking for the link of the `filters` list because that boolean can only be set when(at least) one filter is set.

Reply:
@mike-healey I've changed the code to address your comments